hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -434,10 +434,17 @@ def kill_branch(self) -> NoReturn:
         else:
             trial_data.freeze()
             key = self._cache_key_ir(data=trial_data)
-            try:
-                return self.__data_cache_ir[key]
-            except KeyError:
-                pass
+            if trial_data.status > Status.OVERRUN:
+                try:
+                    return self.__data_cache_ir[key]
+                except KeyError:
+                    pass
+            else:
+                # if we simulated to an overrun, then we our result is certainly
+                # an overrun; no need to consult the cache. (and we store this result
+                # for simulation-less lookup later).
+                self.__data_cache_ir[key] = Overrun
+                return Overrun
```

Comment:
```suggestion
            if trial_data.status > Status.OVERRUN:
                # if we simulated to an overrun, then we our result is certainly
                # an overrun; no need to consult the cache. (and we store this result
                # for simulation-less lookup later).
                self.__data_cache_ir[key] = Overrun
                return Overrun
            try:
                return self.__data_cache_ir[key]
            except KeyError:
                pass
```
Slightly easier to read this way, I think.